### DISCUSSION TITLE
minimum Clang version

### DISCUSSION THREAD

**Discussion by wojdyr** [wjakob/nanobind#713] [General]
The documentation in `docs/index.rst` says "Clang 7+", but I couldn't compile it with Clang < 9.

Clang 7:
```
nanobind/ext/robin_map/include/tsl/robin_hash.h:258:13: error: no member named
      'launder' in namespace 'std'; did you mean 'rand_r'?
    return *std::launder(
```

Clang 8:
```
nanobind/include/nanobind/nb_tuple.h:35:5: error: exception specification of
      explicitly defaulted move constructor does not match the calculated one
    tuple(tuple &&) noexcept = default;
```

For Clang 8, it's enough to remove `noexcept` in two places to make it compile.
Alternatively, the compiler requirement could be changed to "Clang 9+".

**Comment by wjakob** [comment]
For Clang 8, which are the two places where `noexcept` should be removed? (you listed one)

**Reply by wojdyr** [reply]
```diff
diff --git a/include/nanobind/nb_tuple.h b/include/nanobind/nb_tuple.h
index a04b466..ab2c292 100644
--- a/include/nanobind/nb_tuple.h
+++ b/include/nanobind/nb_tuple.h
@@ -32,8 +32,8 @@ template <typename T, typename... Ts> struct tuple<T, Ts...> : tuple<Ts...> {
 
     tuple() = default;
     tuple(const tuple &) = default;
-    tuple(tuple &&) noexcept = default;
-    tuple& operator=(tuple &&) noexcept = default;
+    tuple(tuple &&) = default;
+    tuple& operator=(tuple &&) = default;
     tuple& operator=(const tuple &) = default;
 
     template <typename A, typename... As>
```

**Comment by wjakob** [comment]
Could the `std::launder` issue also be related to the use of an older libstdc++ (which is from GCC) with Clang?

**Reply by wojdyr** [reply]
Probably not, I tested Clang 7, 8 and 9 on the same system.
In https://github.com/abseil/abseil-cpp/pull/1448 std::launder is reported to not work for clang < 8.

**Reply by wjakob** [reply]
I fixed compilation for Clang 8, but Clang 7 will be tricky if ``std::launder`` is broken as this code is part of a dependency. I instead bumped the Clang requirements.
